src/session/session_dhandle.c

```diff
@@ -842,3 +842,31 @@ __wt_session_lock_checkpoint(WT_SESSION_IMPL *session, const char *checkpoint)
     session->dhandle = saved_dhandle;
     return (ret);
 }
+
+/*
+ * __wt_session_check_checkpoint_state --
+ *     Check the state of the checkpoint saved into the session structure - it needs to match the
+ *     dhandle in the current session, and the dhandle that's about to be switched into place.
+ */
+void
+__wt_session_check_checkpoint_state(WT_SESSION_IMPL *session, WT_DATA_HANDLE *new_handle)
+{
+    WT_DATA_HANDLE *old_handle;
+    WT_TXN *txn;
+
+    old_handle = session->dhandle;
+    txn = session->txn;
+
+    if (old_handle != NULL && old_handle->checkpoint != NULL) {
+        /* Ensure checkpoint handles have an associated transaction. */
+        WT_ASSERT(session, F_ISSET(txn, WT_TXN_IS_CHECKPOINT));
+    }
+    if (new_handle->checkpoint != NULL) {
+        /*
+         * If the new handle has a checkpoint, it better match the previous one. TODO: I think the
+         * checkpoint information should travel with the dhandle, not the cursor structure. I don't
+         * think it's OK for a single data handle to service multiple different checkpoints.
+         */
+        WT_ASSERT(session, old_handle != NULL && old_handle->checkpoint != NULL);
+    }
+}
```

Comment:
This, though, I don't think is going to work, if for no other reason than there's a chicken-and-egg problem: checkpoint cursor operations need at some point to load the dhandle with WT_WITH_DHANDLE, and also they load up the checkpoint info with WT_WITH_CHECKPOINT, and one of them has to come before the other. Currently the dhandle gets loaded first.

It might work to change the second assertion to old_handle == NULL || old_handle->checkpoint != NULL; that is, either we're beginning an operation or switching from one checkpoint handle to another.

It is possible that some of the checkpoint information could be moved from the cursor to the dhandle,  but not all of it (specifically, not the transaction...) While each distinct checkpoint of each tree is its own dhandle, tree-level checkpoints can be shared by global checkpoints (this happens anytime the tree is skipped in the checkpoint code) and the transaction snapshot is a property of the global checkpoint. Consequently, if you open different global checkpoints that are the same tree-level checkpoint, they'll share the dhandle. IMO, this is fine; it really is the same tree. You're reading two different views of it. (And if anyone actually does this it seems desirable to share the in-memory pages.) It might make sense in the long run to create some kind of wrapper dhandle that carries the global checkpoint information.

This is hard to reason about in the current code though because global checkpoints barely exist as a thing.

Reply:
Thanks for the response. It took me a while to digest it, but your suggested assert is what I'd wanted (I guess I'm rustier than I thought).
I followed up on the broader question in the JIRA ticket: https://jira.mongodb.org/browse/WT-9344?focusedCommentId=4568043&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4568043